File: tiq-e/strings/6/main.py
from time import perf_counter


class Solution:
    def myAtoi(self, str: str) -> int:
        my_str = str.lstrip()
        if not my_str:
            return 0
        if my_str[0].isdigit():
            return min(self.helper(my_str, 0, len(my_str)), 2147483647)
        if my_str[0] == '+':
            return min(self.helper(my_str, 1, len(my_str)), 2147483647)
        if my_str[0] == '-':
            return max(-self.helper(my_str, 1, len(my_str)), -2147483648)
        return 0

    @staticmethod
    def helper(my_str, p, length):
        res = 0
        while p < length and my_str[p].isdigit():
            res = res * 10 + int(my_str[p])
            p += 1
        return res


s = "42"

start = perf_counter()
result = Solution().myAtoi(s)
end = perf_counter()
print("Duration: " + str(end - start))
print("Result: " + str(result))

File: tiq-e/strings/6/main.rs
use std::time::Instant;

struct Solution;

impl Solution {
    pub fn my_atoi(str: String) -> i32 {
        let (n, s) = match str.chars().skip_while(|x| x.is_whitespace()).take(1).next() {
            Some('+') => (1, 1),
            Some(x) if x.is_digit(10) => (0, 1),
            Some('-') => (1, -1),
            _ => return 0,
        };
        let mut res = 0i32;
        let overflow = if s > 0 { std::i32::MAX } else { std::i32::MIN };
        for c in str
            .chars()
            .skip_while(|x| x.is_whitespace())
            .skip(n)
            .take_while(|x| x.is_digit(10))
        {
            let (r, o) = res.overflowing_mul(10);
            if o {
                return overflow;
            }
            let (r, o) = r.overflowing_add(s * (c as i32 - '0' as i32));
            if o {
                return overflow;
            }
            res = r;
        }
        res
    }
}

fn main() {
    let s = "42".to_string();

    let start = Instant::now();
    let result = Solution::my_atoi(s);
    let end = start.elapsed().as_micros();
    println!("Duration: {:?}", end);
    println!("Result: {:?}", result);
}
